rust/src/x509/mod.rs

```diff
@@ -111,40 +112,47 @@ pub unsafe extern "C" fn SCX509GetSubjectAltNameLen(ptr: *const X509) -> u16 {
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn SCX509GetSubjectAltNameAt(ptr: *const X509, idx: u16) -> *mut c_char {
+pub unsafe extern "C" fn SCX509GetSubjectAltNameAt(ptr: *const X509, idx: u16, san: *mut *mut u8, san_len: *mut u32) {
     if ptr.is_null() {
-        return std::ptr::null_mut();
+        *san_len = 0;
+        *san = std::ptr::null_mut();
     }
     let x509 = cast_pointer! {ptr, X509};
     let san_list = x509.0.tbs_certificate.subject_alternative_name();
     if let Ok(Some(sans)) = san_list {
         let general_name = &sans.value.general_names[idx as usize];
         let dns_name = SCGeneralName(general_name);
-        return rust_string_to_c(dns_name.to_string());
+        let dn = dns_name.to_string().into_bytes();
+        *san_len = dn.len() as u32;
+        *san = Box::into_raw(dn.into_boxed_slice()) as *mut u8;
     }
-    return std::ptr::null_mut();
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn SCX509GetIssuer(ptr: *const X509) -> *mut c_char {
+pub unsafe extern "C" fn SCX509GetIssuer(ptr: *const X509, issuer_name: *mut *mut u8, issuer_len: *mut u32) {
     if ptr.is_null() {
-        return std::ptr::null_mut();
+        *issuer_len = 0;
+        *issuer_name = std::ptr::null_mut();
     }
     let x509 = cast_pointer! {ptr, X509};
-    let issuer = x509.0.tbs_certificate.issuer.to_string();
-    rust_string_to_c(issuer)
+    let issuer = x509.0.tbs_certificate.issuer.to_string().into_bytes();
+
+    *issuer_len = issuer.len() as u32;
+    *issuer_name = Box::into_raw(issuer.into_boxed_slice()) as *mut u8;
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn SCX509GetSerial(ptr: *const X509) -> *mut c_char {
+pub unsafe extern "C" fn SCX509GetSerial(ptr: *const X509, serial_num: *mut *mut u8, serial_len: *mut u32) {
     if ptr.is_null() {
-        return std::ptr::null_mut();
+        *serial_len = 0;
+        *serial_num = std::ptr::null_mut();
     }
     let x509 = cast_pointer! {ptr, X509};
     let raw_serial = x509.0.tbs_certificate.raw_serial();
     let v: Vec<_> = raw_serial.iter().map(|x| format!("{:02X}", x)).collect();
     let serial = v.join(":");
-    rust_string_to_c(serial)
+    *serial_len = serial.len() as u32;
+    *serial_num = Box::into_raw(serial.into_bytes().into_boxed_slice()) as *mut u8;
```

Comment:
do we need to box it if the caller is supposed to copy it?

Reply:
No. I think I was at some point just trying to make it compile. This follows a bad pattern of double alloc anyway pointed by Philippe in an earlier rev. So, removing the allocs on the C side and then we need the Box. Coming up in the next rev.